### ISSUE TITLE
Package TRegex for regular Java program consumption

### ISSUE THREAD

**Issue by rednaxelafx** [oracle/graal#709]
The TRegex looks like a very nice regular expression engine implementation. Thanks a lot to the Graal/Truffle team for making it available!

It would be a pity to be confined to just the Truffle languages' world. It'd be really cool to allow regular Java programs to be able to use it as well.

A few features that would be nice-to-have in this scenario would be:
- Accept Java regular expression syntax (to the extent that the DFA can support)
- Wrap TRegex into a familiar `Pattern` / `Matcher` API similar to the one in the JDK
- Make package available on popular package managers (e.g. Maven?)
- Maybe compile the package into SVM-style native image to reduce initial overhead?

The API wrapper was done in an NFA-based [RE2/J](https://github.com/google/re2j) as well. RE2/J doesn't support all features that `java.util.regex` does, but by wrapping the engine into a familiar API, it eases the migration, so that regular expressions that happen to be supported can switch from using the JDK one to RE2/J by simply changing the import statements.

**Comment by wirthi** [comment]
Hi,

we currently have a student investigating that area. We might be able to publish that as an experimental feature (at least your points #1, #2 and #4; not sure about #3). 

Maybe @djoooooe can give you a quick update here once this is merged.

Christian

**Comment by rednaxelafx** [comment]
Great news! Thank you @wirthi and @djoooooe for picking up this idea, still looking forward to having a fast DFA RE impl for Java ^_^
